packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -166,6 +136,44 @@ export function makeWatchedPromiseManager(
     }
   }
 
+  function watchPromise(p, watcher, ...args) {
+    // The following wrapping defers setting up the promise watcher itself to a
+    // later turn so that if the promise to be watched was the return value from
+    // a preceding eventual message send, then the assignment of a vpid to that
+    // promise, which happens in a turn after the initiation of the send, will
+    // have happened by the time the code below executes, and thus when we call
+    // `convertValToSlot` on the promise here we'll get back the vpid that was
+    // assigned rather than generating a new one that nobody knows about.
+    Promise.resolve().then(() => {
```

Comment:
Yeah, we're a little nervous too.  The problem is that there _isn't_ already a slot allocated so `convertValToSlot` allocates one, but then the message send code allocates another one (in a way that's entirely different).  There's magic voodoo promise plumbing in the middle that one interferes with at one's peril.

Reply:
But by the time `Promise.resolve()` reaction runs and calls convertValToSlot, the goal is that it would be allocated then, or did I misunderstood? If that's the case, there is no way to assert it?

Reply:
Ah, I misunderstood what you were saying.  Yes, by the time `Promise.resolve()` runs everything is OK -- so I gather you're suggesting an option to assert that everything actually _is_ OK at that point?

Reply:
Aye, just in case something (e.g. the promise implementation) changes from under our feet in some conditions, and our assumption no longer holds

Reply:
Well, if there _is_ a `slotToVal` entry then `convertValToSlot` will just use it, so in that case there's not actually extra machinery in `convertValToSlot` that we're leaning on here.  And in that case we might just call `requiredValForSlot`.  But I'm pretty sure this would not do the right thing in the case where the vpid is not being generated by a message send, in which case there might well not be an existing `slotToVal` entry and we really do need `convertValToSlot` to do all the work.

Reply:
And never mind that last comment.  I'm always getting `convertSlotToVal` and `convertValToSlot` mixed up, because we also have `getValForSlot` and `getSlotForVal` where the pattern of the names is reversed.  This confusion has been on our hitlist for a while now.

Reply:
So let me try again: in the case where the vpid is being generated by a message send, we need the deferred execution that `Promise.resolve()...` provides so that `convertValToSlot` will see the vpid that the message send produced.  However, in the case where it's not a message send, we really do need `convertSlotToVal` to generate the vpid.  In the latter case we don't need the deferral, but at the point where `watchPromise` is called we don't know which case we're in, and the deferral is harmless in the case where we don't actually need it.